src/txmempool.cpp

```diff
@@ -151,33 +152,47 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
+                                           const uint64_t limitAncestorCount,
+                                           const uint64_t limitAncestorSize,
+                                           const uint64_t limitDescendantCount,
+                                           const uint64_t limitDescendantSize,
+                                           std::string &errString,
+                                           const bool fSearchForParents /*=true*/) const
 {
-    CTxMemPoolEntry::Parents staged_ancestors;
-    const CTransaction &tx = entry.GetTx();
+    const size_t total_virtual_size = std::accumulate(entries.cbegin(), entries.cend(), 0,
+                                      [](int64_t sum, const auto& entry)
+                                      { return sum + GetVirtualTransactionSize(entry.get().GetTx()); });
+    const size_t total_count = entries.size();
 
+    CTxMemPoolEntry::Parents staged_ancestors;
     if (fSearchForParents) {
-        // Get parents of this transaction that are in the mempool
-        // GetMemPoolParents() is only valid for entries in the mempool, so we
-        // iterate mapTx to find parents.
-        for (unsigned int i = 0; i < tx.vin.size(); i++) {
-            std::optional<txiter> piter = GetIter(tx.vin[i].prevout.hash);
-            if (piter) {
-                staged_ancestors.insert(**piter);
-                if (staged_ancestors.size() + 1 > limitAncestorCount) {
-                    errString = strprintf("too many unconfirmed parents [limit: %u]", limitAncestorCount);
-                    return false;
+        for (const auto& entry : entries) {
+            // Get parents of this transaction that are in the mempool
+            // GetMemPoolParents() is only valid for entries in the mempool, so we
+            // iterate mapTx to find parents.
+            for (const auto& input : entry.get().GetTx().vin) {
+                std::optional<txiter> piter = GetIter(input.prevout.hash);
+                if (piter) {
+                    staged_ancestors.insert(**piter);
+                    if (staged_ancestors.size() + total_count > limitAncestorCount) {
+                        errString = strprintf("too many unconfirmed parents [limit: %u]", limitAncestorCount);
+                        return false;
+                    }
                 }
             }
         }
     } else {
-        // If we're not searching for parents, we require this to be an
-        // entry in the mempool already.
-        txiter it = mapTx.iterator_to(entry);
-        staged_ancestors = it->GetMemPoolParentsConst();
+        // If we're not searching for parents, we require all entries to be in the mempool already.
+        // We should never use CalculateMemPoolAncestors on a set of transactions that are already
+        // in the mempool.
+        assert(entries.size() == 1);
+        txiter it = mapTx.iterator_to(entries[0].get());
+        staged_ancestors = it->GetMemPoolParents();
```

Comment:
What happened to the `Const` here? This used to be `GetMemPoolParentsConst()`.

Reply:
Fixed